File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  checkItemName(itemName: string, nameToCompare: string){
    return itemName != nameToCompare;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) {
      let itemName =this.items[i].name;
      if (this.checkItemName(itemName, "Aged Brie") && this.checkItemName(itemName, 'Backstage passes to a TAFKAL80ETC concert')) {
        if (this.items[i].quality > 0) {
          if (this.checkItemName(itemName, 'Sulfuras, Hand of Ragnaros')) {
            this.items[i].quality = this.items[i].quality - 1
          }
        }
      } else {
        if (this.items[i].quality < 50) {
          this.items[i].quality = this.items[i].quality + 1
          if (!this.checkItemName(itemName, 'Backstage passes to a TAFKAL80ETC concert')) {
            if (this.items[i].sellIn < 11) {
              if (this.items[i].quality < 50) {
                this.items[i].quality = this.items[i].quality + 1
              }
            }
            if (this.items[i].sellIn < 6) {
              if (this.items[i].quality < 50) {
                this.items[i].quality = this.items[i].quality + 1
              }
            }
          }
        }
      }
      if (this.checkItemName(itemName, 'Sulfuras, Hand of Ragnaros')) {
        this.items[i].sellIn = this.items[i].sellIn - 1;
      }
      if (this.items[i].sellIn < 0) {
        if (this.checkItemName(itemName, 'Aged Brie') && this.checkItemName(itemName, 'Backstage passes to a TAFKAL80ETC concert')) {
            if (this.items[i].quality > 0) {
              if (this.checkItemName(itemName, 'Sulfuras, Hand of Ragnaros')) {
                this.items[i].quality = this.items[i].quality - 1
              }
          } else {
            this.items[i].quality = this.items[i].quality - this.items[i].quality
          }
        } else {
          if (this.items[i].quality < 50) {
            this.items[i].quality = this.items[i].quality + 1
          }
        }
      }
    }

    return this.items;
  }
}

File: python/gilded_rose.py
# -*- coding: utf-8 -*-

class GildedRose(object):

    def __init__(self, items):
        self.items = items

    def update_quality(self):
        for item in self.items:
            if self.is_not_Backstage_pass_and_not_aged_brie(item.name):
                if item.quality > 0:
                    if self.is_not_Sulfuras_Hand_of_Ragnaros(item.name):
                        item.quality = item.quality - 1
            else:
                if item.quality < 50:
                    item.quality = item.quality + 1
                    if self.is_Backstage_pass_less_than_11_days(item.name,item.sell_in,item.quality):
                        item.quality = item.quality + 1
                    if self.is_Backstage_pass_less_than_6_days(item.name,item.sell_in,item.quality):
                        item.quality = item.quality + 1
            if self.is_not_Sulfuras_Hand_of_Ragnaros(item.name):
                item.sell_in = item.sell_in - 1
            if item.sell_in < 0:
                if item.name != "Aged Brie":
                    if item.name != "Backstage passes to a TAFKAL80ETC concert":
                        if item.quality > 0:
                            if self.is_not_Sulfuras_Hand_of_Ragnaros(item.name):
                                item.quality = item.quality - 1
                    else:
                        item.quality = item.quality - item.quality
                else:
                    if item.quality < 50:
                        item.quality = item.quality + 1

    def is_Backstage_pass_less_than_11_days(self,item_name, item_sell_in ,item_quality):
        return item_sell_in < 11 and item_quality < 50 and item_name =="Backstage passes to a TAFKAL80ETC concert"

    def is_Backstage_pass_less_than_6_days(self,item_name, item_sell_in ,item_quality):
        return item_sell_in < 6 and item_quality < 50 and item_name =="Backstage passes to a TAFKAL80ETC concert"

    def is_not_Backstage_pass_and_not_aged_brie(self,item_name):
        return item_name != "Aged Brie" and item_name != "Backstage passes to a TAFKAL80ETC concert"

    def is_not_Sulfuras_Hand_of_Ragnaros(self,item_name):
        return item_name!= "Sulfuras, Hand of Ragnaros"


class Item:
    def __init__(self, name, sell_in, quality):
        self.name = name
        self.sell_in = sell_in
        self.quality = quality

    def __repr__(self):
        return "%s, %s, %s" % (self.name, self.sell_in, self.quality)
